# nf-core/sangersomatic: Citations

## [nf-core](https://pubmed.ncbi.nlm.nih.gov/32055031/)

> Ewels PA, Peltzer A, Fillinger S, Patel H, Alneberg J, Wilm A, Garcia MU, Di Tommaso P, Nahnsen S. The nf-core framework for community-curated bioinformatics pipelines. Nat Biotechnol. 2020 Mar;38(3):276-278. doi: 10.1038/s41587-020-0439-x. PubMed PMID: 32055031.

## [Nextflow](https://pubmed.ncbi.nlm.nih.gov/28398311/)

> Di Tommaso P, Chatzou M, Floden EW, Barja PP, Palumbo E, Notredame C. Nextflow enables reproducible computational workflows. Nat Biotechnol. 2017 Apr 11;35(4):316-319. doi: 10.1038/nbt.3820. PubMed PMID: 28398311.

## Pipeline tools

- [gatk](https://genome.cshlp.org/content/20/9/1297)

> McKenna A, Hanna M, Banks E, Sivachenko A, Cibulskis K, Kernytsky A, Garimella K, Altshuler D, Gabriel S, Daly M, DePristo MA. (2010). The Genome Analysis Toolkit: a MapReduce framework for analyzing next-generation DNA sequencing data. Genome Res, 20:1297-303. DOI: 10.1101/gr.107524.110.

- [hairpin](https://www.nature.com/articles/s41596-020-00437-6)

> Ellis, P., Moore, L., Sanders, M.A. et al. Reliable detection of somatic mutations in solid tissues by laser-capture microdissection and low-input DNA sequencing. Nat Protoc 16, 841–871 (2021). https://doi.org/10.1038/s41596-020-00437-6

> CASM IT, Sanger institute, tools are [here](https://github.com/cancerit/hairpin-wrapper)

- [vafCorrect](https://github.com/cancerit/vafCorrect)

> CASM IT, Sanger institute

- [SigProfilerMatrixGenerator](https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-019-6041-2)

> Bergstrom EN, Huang MN, Mahto U, Barnes M, Stratton MR, Rozen SG, and Alexandrov LB (2019) SigProfilerMatrixGenerator: a tool for visualizing and exploring patterns of small mutational events. BMC Genomics 20, Article number: 685.

- [SigProfilerPlotting](https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-019-6041-2)

> Bergstrom EN, Huang MN, Mahto U, Barnes M, Stratton MR, Rozen SG, and Alexandrov LB (2019) SigProfilerMatrixGenerator: a tool for visualizing and exploring patterns of small mutational events. BMC Genomics 20, Article number: 685.

- [MPBoot](https://bmcecolevol.biomedcentral.com/articles/10.1186/s12862-018-1131-3)

> Hoang, D.T., Vinh, L.S., Flouri, T. et al. MPBoot: fast phylogenetic maximum parsimony tree inference and bootstrap approximation. BMC Evol Biol 18, 11 (2018). https://doi.org/10.1186/s12862-018-1131-3

- [Sequoia](https://www.nature.com/articles/s41596-024-00962-8)

> Coorens, T.H.H., Spencer Chapman, M., Williams, N. et al. Reconstructing phylogenetic trees from genome-wide somatic mutations in clonal samples. Nat Protoc 19, 1866–1886 (2024). https://doi.org/10.1038/s41596-024-00962-8

- [treemut](https://github.com/NickWilliamsSanger/treemut)

> Nicholas Williams, Sanger institute

## Software packaging/containerisation tools


- [Docker](https://dl.acm.org/doi/10.5555/2600239.2600241)

  > Merkel, D. (2014). Docker: lightweight linux containers for consistent development and deployment. Linux Journal, 2014(239), 2. doi: 10.5555/2600239.2600241.

- [Singularity](https://pubmed.ncbi.nlm.nih.gov/28494014/)

  > Kurtzer GM, Sochat V, Bauer MW. Singularity: Scientific containers for mobility of compute. PLoS One. 2017 May 11;12(5):e0177459. doi: 10.1371/journal.pone.0177459. eCollection 2017. PubMed PMID: 28494014; PubMed Central PMCID: PMC5426675.
